testutil/fuse/fuse.go

```diff
@@ -0,0 +1,164 @@
+// +build linux darwin
+
+// Copyright 2019 The qiffang Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"fmt"
+	"path/filepath"
+	"runtime"
+	"syscall"
+	"time"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+)
+
+type loopBackFs struct {
+	pathfs.FileSystem
+}
+
+// hookFs filesystem that shunts all request to an underlying file system.
+// Its main purpose is to provide test coverage.
+type hookFs struct {
+	original   string
+	mountpoint string
+	fsName     string
+	loopBackFs
+	hook Hook
+}
+
+// String implements hanwen/go-fuse/fuse/pathfs.FileSystem. You are not expected to call h manually.
+func (h *hookFs) String() string {
+	return fmt.Sprintf("HookFs{Original=%s, Mountpoint=%s, FsName=%s, Underlying fs=%s, hook=%s}",
+		h.original, h.mountpoint, h.fsName, h.loopBackFs.String(), h.hook)
+}
+
+// Rename implements hanwen/go-fuse/fuse/pathfs.FileSystem. You are not expected to call h manually.
```

Comment:
It is most common to note that a given struct implements an interface and the describe the actual funcs closer to what they actually do.

```suggestion
// Rename calls the injected rename hooks if those exist or the underlying loopback fs Rename api.
```

Reply:
Got it, Done